pkg/manifests/manifests.go

```diff
@@ -3896,21 +3903,31 @@ func (f *Factory) TelemeterClientServiceAccount() (*v1.ServiceAccount, error) {
 	return s, nil
 }
 
+func (f *Factory) TelemeterClientSecretTemplate() (*v1.Secret, error) {
+	return f.NewSecret(f.assets.MustNewAssetReader(TelemeterClientSecret))
+}
+
 // TelemeterClientSecret generates a new Secret for Telemeter client.
-func (f *Factory) TelemeterClientSecret() (*v1.Secret, error) {
-	s, err := f.NewSecret(f.assets.MustNewAssetReader(TelemeterClientSecret))
+func (f *Factory) TelemeterClientSecret(secretData map[string][]byte) (*v1.Secret, error) {
+	s, err := f.TelemeterClientSecretTemplate()
 	if err != nil {
 		return nil, err
 	}
 
+	if f.config.ClusterMonitoringConfiguration.TelemeterClientConfig.Token != "" {
+		s.Data["token"] = []byte(f.config.ClusterMonitoringConfiguration.TelemeterClientConfig.Token)
+	}
+
 	salt, err := GeneratePassword(32)
 	if err != nil {
 		return nil, fmt.Errorf("failed to generate Telemeter client salt: %v", err)
 	}
 	s.Data["salt"] = []byte(salt)
 
-	if f.config.ClusterMonitoringConfiguration.TelemeterClientConfig.Token != "" {
-		s.Data["token"] = []byte(f.config.ClusterMonitoringConfiguration.TelemeterClientConfig.Token)
+	oldToken := string(secretData["token"])
+	oldSalt, saltExists := secretData["salt"]
+	if f.config.ClusterMonitoringConfiguration.TelemeterClientConfig.Token != oldToken && saltExists {
```

Comment:
This is using the old salt when the token is updated, but what we want is "Only update the salt
in the secret when the token is updated", that is:

- if the token is updated then use the new salt: with this code `f.config.ClusterMonitoringConfiguration.TelemeterClientConfig.Token != oldToken && saltExists` would be `true`, so the old salt would override the new one
- if the token is not updated then use the old salt: with this code `f.config.ClusterMonitoringConfiguration.TelemeterClientConfig.Token != oldToken && saltExists` would be `false`, so we would use the previous assignment `s.Data["salt"] = []byte(salt)` that employs the new salt

Can we add a unit test for the salt update logic?

Reply:
I agree with Juan, the update logic seems off.
:+1: if we can add a test for this.

Reply:
I've added tests 👌